extensions/ql-vscode/src/data-extensions-editor/extension-packs.ts

```diff
@@ -66,27 +73,107 @@ export async function pickModelFile(
     }
   }
 
-  const fileOptions: Array<{ label: string; file: string }> = [];
+  const fileOptions: Array<{ label: string; file: string | null }> = [];
   for (const file of modelFiles) {
     fileOptions.push({
       label: relative(extensionPackPath, file),
       file,
     });
   }
+  fileOptions.push({
+    label: "Create new model file",
+    file: null,
+  });
 
   progress({
     message: "Choosing model file...",
     step: 3,
     maxStep: 3,
   });
 
-  const fileOption = await window.showQuickPick(fileOptions, {
-    title: "Select model file to use",
-  });
+  const fileOption = await window.showQuickPick(
+    fileOptions,
+    {
+      title: "Select model file to use",
+    },
+    token,
+  );
 
   if (!fileOption) {
-    return;
+    return undefined;
+  }
+
+  if (fileOption.file) {
+    return fileOption.file;
+  }
+
+  return pickNewModelFile(token, databaseItem, extensionPackPath);
+}
+
+async function pickNewModelFile(
+  token: CancellationToken,
+  databaseItem: DatabaseItem,
+  extensionPackPath: string,
+) {
+  const qlpackPath = await getQlPackPath(extensionPackPath);
+  if (!qlpackPath) {
+    void showAndLogErrorMessage(
+      `Could not find any of ${QLPACK_FILENAMES.join(
+        ", ",
+      )} in ${extensionPackPath}`,
+    );
+    return undefined;
+  }
+
+  const qlpack = await loadYaml(await readFile(qlpackPath, "utf8"), {
+    filename: qlpackPath,
+  });
+  if (typeof qlpack !== "object" || qlpack === null) {
+    void showAndLogErrorMessage(`Could not parse ${qlpackPath}`);
+    return undefined;
+  }
+
+  const dataExtensionPatterns = qlpack.dataExtensions ?? [];
+  if (!Array.isArray(dataExtensionPatterns)) {
+    void showAndLogErrorMessage(
+      `Expected 'dataExtensions' to be an array in ${qlpackPath}`,
+    );
+    return undefined;
+  }
+
+  const filename = await window.showInputBox(
+    {
+      title: "Enter the name of the new model file",
+      value: `models/${databaseItem.name.replaceAll("/", ".")}.model.yml`,
+      validateInput: async (value: string): Promise<string | undefined> => {
+        const path = resolve(extensionPackPath, value);
+
+        if (await pathExists(path)) {
+          return "File already exists";
+        }
+
+        const isInExtensionPack = !relative(extensionPackPath, path).startsWith(
+          "..",
+        );
+        if (!isInExtensionPack) {
```

Comment:
I'm finding this double negative hard to grok. What do you think of this instead?

```suggestion
        const notInExtensionPack = relative(extensionPackPath, path).startsWith(
          "..",
        );
        if (notInExtensionPack) {
```